README.md
# Best practices for constructing, preparing, and evaluating protein-ligand binding affinity benchmarks

## Abstract

Free energy calculations are rapidly becoming indispensable in structure-enabled drug discovery programs. As new methods, force fields, and implementations are developed, assessing their expected accuracy on real-world systems (benchmarking) becomes critical to provide users with an assessment of the accuracy expected when these methods are applied within their domain of applicability, and developers with a way to assess the expected impact of new methodologies. These assessments require construction of a benchmark—a set of well-prepared, high quality systems with corresponding experimental measurements designed to ensure the resulting calculations provide a realistic assessment of expected performance when these methods are deployed within their domains of applicability. To date, the community has not yet adopted a common standardized benchmark, and existing benchmark reports suffer from a myriad of issues, including poor data quality, limited statistical power, and statistically deficient analyses, all of which can conspire to produce benchmarks that are poorly predictive of real-world performance. Here, we address these issues by presenting guidelines for 

1. curating experimental data to develop meaningful benchmark sets, 
2. preparing benchmark inputs according to best practices to facilitate widespread adoption, and 
3. analysis of the resulting predictions to enable statistically meaningful comparisons among methods and force fields. 
 
We highlight challenges and open questions that remain to be solved in these areas, as well as recommendations for the collection of new datasets that might optimally serve to measure progress as methods become systematically more reliable. Finally, we provide a curated, versioned, open, standardized benchmark set adherent to these standards [protein-ligand-benchmark](https://github.com/openforcefield/protein-ligand-benchmark)  and an open source toolkit for implementing standardized best practices assessments [arsenic](https://github.com/openforcefield/Arsenic) for the community to use as a standardized assessment tool. While our main focus is free energy methods based on molecular simulations, these guidelines should prove useful for assessment of the rapidly growing field of machine learning methods for affinity prediction as well.

## List of Authors

- DH: David Hahn   
- CIB: Christopher I. Bayly
- HBM: Hannah Bruce Macdonald
- JDC: John D. Chodera
- VG: Vytautas Gapsys
- ASJSM: Antonia S. J. S. Mey
- DLM:  David L. Mobley
- LPB: Laura Perez Benito
- CS Christina E.M. Schindler
- GT: Gary Tresadern
- GW: Gregory L. Warren

## List of Contributors
(non-author list of people who contributed to document)

## Paper Writing as Code Development
<!-- This discussion is so that people know how to contribute to your document. -->
This paper is being developed as a living document, open to changes from the community.
You can read more about the concept of writing a paper in the same way one would write software code in the essay ["Paper writing as code development"](https://livecomsjournal.github.io/paper_code.html).
If you have comments or suggestions, we welcome them! Please [submit them as issues](https://guides.github.com/features/issues/) to this GitHub repository so they can be recorded and given credit for the contribution.
Specific changes can be proposed [via pull requests](https://help.github.com/articles/about-pull-requests/).


## Online Resources
Original brainstorming document: (https://docs.google.com/document/d/1lCGcol6jYLQmcfqrUv9h_FsWygTZzqYxqgjOLCyMoL4/edit)

## Design Notes:
- Figure labels are in captital bold face
- Standard font for all figures should be Helvetica. Font size for lables should be at least the size of the document text. 

## List of Released Versions
- revision for version 1.0


CHANGELOG
v 1.0 
-----
- Accepted version to LiveComms 


examples/structure_analysis/Score_pdb.sh
#!/bin/bash
# Prepare pdb files using OpenEye Spruce
# """
#
# Pipeline Step: Prepare raw PDB protein structures and score them.
#
# Description: This script pulls pdb files from a list and scores them using the Iridium criteria.
#
# Step Flowchart:
#     * Generate a text file with a pdb id per line for the structures to be scored.
#     * The inputs are the name of the pdb id text file and the Target name.
#     * Select which quality levels to structures to use:
#          * HT: Highly Trustworthy structures - a good model for the data and little or no missing density
#          * ALL: This include HT and MT (Moderarely Trustworthy) but no NT (Not Trustworthy) structures
#     * Output is a .csv file named target_HT/ALL_scoring_list.csv
# """

# This script requires installation of the
# OpenEye application Spruce, python OEChem toolkit and docopt python package

# Test for input command line arguments
if [ $# -ne 3 ]; then
    echo "Please supply: <pdb list file name> <target name> <Score Type (HT ALL)>"
    exit
fi
# Constants
zero=0
DPI_limit=0.75
LaD_limit=1.1
ASaD_limit=0.49
Iridium_score=0
Final_Iridium_score=1000000
HT_flag=0
MT_flag=0
target_name_count=$(tr -dc '_' <<<"$2" | wc -c)
(( final=target_name_count+1 ))

# Test for pdb data directory
# Create a pdb_data directory
# for downloading data into
if [ ! -d pdb_data ]
then
    mkdir pdb_data
    cd pdb_data || exit
else
    cd pdb_data || exit
    mkdir backup
    mv -- *.* backup/
fi

for file in $(cat ../"$1")
do
    getstructure "$file"
done
cd ../

# Test if there is a prep directory
# Create if not
if [ ! -d prep ]
then
    mkdir prep
    cd prep || return
else
    cd prep || return
    mkdir backup
    mv -- *.* backup/
fi

# Run spruce once to assess the quality of the structures
# Choosing to use .cif files versus pdb
# cif should have ligand bond order information while pdb does not
for file in ../pdb_data/*.cif
do
    tmp=$(echo "$file" | cut -d / -f 3)
    pdb_prefix=$(echo "$tmp" | cut -d . -f 1)
    map_prefix=$(echo "$file" | cut -d . -f 1-3)
    for list in "$map_prefix".m*
    do
        if [ -e "$list" ]
        then
            spruce -in "$file" -map "$list" -prefix "$2"-"$pdb_prefix"
        else
            spruce -in "$file" -prefix "$2"-"$pdb_prefix"
        fi
    done
done

# Determine the best structure based on
# ligand density, active site density, and DPI
# for this score LOWER is BETTER

selection="$3"

# print title for csv output file
if [[ "$selection" == *HT* ]]
then
    if [ -e ../"$2"\_HT_scoring_list.csv ]; then rm ../"$2"\_HT_scoring_list.csv; fi
    echo "PDB id,Ligand,Iridium Category,Ligand density,Active-site density,DPI,Iridium score,Packing residues" >> ../"$2"\_HT_scoring_list.csv
elif [[ "$selection" == *A[lL][lL]* ]]
then
    if [ -e ../"$2"\_ALL_scoring_list.csv ]; then rm ../"$2"\_ALL_scoring_list.csv; fi
    echo "PDB id,Ligand,Iridium Category,Ligand density,Active-site density,DPI,Iridium score,Packing residues" >> ../"$2"\_ALL_scoring_list.csv
else
    echo "Please select either HT or ALL (case sensitive)"
    exit
fi

HT_test=$(grep "Iridium" *output.log | grep -c "HT")
MT_test=$(grep "Iridium" *output.log | grep -c "MT")

if [ "$HT_test" -gt 0 ] || [ "$MT_test" -gt 0 ]
then
    for file in $(ls -1 *_output.log | cut -d \_ -f 1-"$final")
    do
        category=$(grep "Iridium" "$file"\_output.log | cut -d : -f 3 | cut -d , -f 1)
        LaD=$(grep "Iridium" "$file"\_output.log | cut -d : -f 4 | cut -d , -f 1)
        ASaD=$(grep "Iridium" "$file"\_output.log | cut -d : -f 5 | cut -d , -f 1)
        DPI=$(grep "Iridium" "$file"\_output.log | cut -d : -f 6 | cut -d , -f 1)
        file_name=$(grep "Iridium" "$file"\_output.log | cut -d : -f 2 | cut -d \> -f 1)
        ligand_name=$(grep "Iridium" "$file"\_output.log | cut -d : -f 2 | cut -d \> -f 2 | cut -d . -f 1 | cut -d , -f 1)
        packing_name=$(grep "Iridium" "$file"\_output.log | cut -d : -f 10 | cut -d , -f 1)
        h=0
        i=0
        j=0
        k=0
        l=0
        m=0
        n=0
        o=0
        C_tmp=""
        L_tmp="-1"
        A_tmp="-1"
        D_tmp="100"
        F_tmp=""
        Lig_tmp=""
        #P_tmp=""
        for job in ${LaD[*]}
        do
            if (( $(echo "$job > $L_tmp" |bc -l) )) && (( $(echo "$job <= 1.0" | bc -l) ))
            then
                L_tmp="$job"
                (( i=i+1 ))
            fi
        done
        for job in ${ASaD[*]}
        do
            if (( $(echo "$job > $A_tmp" |bc -l) )) && (( $(echo "$job <= 1.0" | bc -l) ))
            then
                A_tmp="$job"
                (( j=j+1 ))
            fi
        done
        for job in ${DPI[*]}
        do
            if (( $(echo "$job < $D_tmp" |bc -l) ))
            then
                D_tmp="$job"
                (( k=k+1 ))
            fi
        done
        for job in ${file_name[*]}
        do
            (( l=l+1 ))
            #echo $job $l $i $j $k
            if [ "$l" -eq "$i" ] && [ "$l" -eq  "$j" ] && [ "$l" -eq  "$k" ]
            then
                F_tmp="$job"
                n="$l"
            elif [ "$l" -eq  "$i" ] &&  [ "$l" -eq  "$k" ]
            then
                F_tmp="$job"
                n="$l"
            elif [ "$l" -eq  "$i" ] && [ "$l" -eq  "$j" ]
            then
                F_tmp="$job"
                n="$l"
            elif [ "$l" -eq  "$i" ]
            then
                F_tmp="$job"
                n="$l"
           fi
        done
        for job in ${category[*]}
        do
            (( h=h+1 ))
            #echo $i $j $k $job
            if [ "$h" -eq "$n" ]
            then
                C_tmp="$job"
            fi
        done
        for job in ${ligand_name[*]}
        do
            (( m=m+1 ))
            if [ "$m" -eq "$n" ]
            then
                Lig_tmp="$job"
            fi
        done
        for job in ${packing_name[*]}
        do
            (( o=o+1 ))
            if [ "$o" -eq "$n" ]
            then
                packing="$job"
            fi
        done
        LaD="$L_tmp"
        ASaD="$A_tmp"
        if (( $(echo "$D_tmp < $DPI_limit" | bc -l) ))
        then
            DPI="$D_tmp"
        else
            factor=$(echo 0.75 / "$D_tmp" | bc -l)
            factor=$(echo "$factor" / 101 | bc -l)
            DPI=$(echo "$DPI_limit" - "$factor" | bc -l)
            #echo $DPI
        fi
        LaD=$(echo "$LaD_limit" - "$LaD" | bc -l)
        ASaD=$(echo "$ASaD" - "$ASaD_limit" | bc -l)
        DPI=$(echo "$DPI_limit" - "$DPI" | bc -l)
        Numerator=$(echo "$LaD" / "$ASaD" | bc -l)
        Iridium_score=$(echo "$Numerator" / "$DPI" | bc -l)
        Iridium_cat="$C_tmp"

        if [[ "$selection" == HT && "$Iridium_cat" == *HT* ]]
        then
            echo "$F_tmp","$Lig_tmp","$Iridium_cat","$L_tmp","$A_tmp","$D_tmp","$Iridium_score","$packing" >> ../"$2"\_HT_scoring_list.csv
            if (( $(echo "$Iridium_score < $Final_Iridium_score" | bc -l) ))
            then
                Final_Iridium_score="$Iridium_score"
                Best_structure="$F_tmp"
                Best_ligand="$Lig_tmp"
            fi
            HT_flag=1
        fi

        if [[ "$selection" == A[lL][lL]  && "$Iridium_cat" == *[HM]T* ]]
        then
            echo "$F_tmp","$Lig_tmp","$Iridium_cat","$L_tmp","$A_tmp","$D_tmp","$Iridium_score","$packing" >> ../"$2"\_ALL_scoring_list.csv
            if (( $(echo "$Iridium_score < $Final_Iridium_score" | bc -l) ))
            then
                Final_Iridium_score="$Iridium_score"
                Best_structure="$F_tmp"
                Best_ligand="$Lig_tmp"
            fi
            MT_flag=1
        fi
    done
else
    echo "No good structures in this set"
    exit
fi

if [[ "$selection" == HT && "$HT_flag" -eq 1 ]]
then
    echo ""
    echo "The best structure is " "$Best_structure" "$Best_ligand" "$Final_Iridium_score" >> ../"$2"\_HT_scoring_list.csv
    uniq ../"$2"\_HT_scoring_list.csv tmp.csv
    mv tmp.csv ../"$2"\_HT_scoring_list.csv
elif [[ "$selection" == A[lL][lL] && "$MT_flag" -eq 1 ]]
then
    echo ""
    echo "The best structure is " "$Best_structure" "$Best_ligand" "$Final_Iridium_score" >> ../"$2"\_ALL_scoring_list.csv
    uniq ../"$2"\_ALL_scoring_list.csv tmp.csv
    mv tmp.csv ../"$2"\_ALL_scoring_list.csv
else
    if [[ "$HT_flag" -ne 1 && "$selection" == HT ]]
    then
        rm ../"$2"\_HT_scoring_list.csv
        echo "There are no HT structures in this data set"
    fi
    if [[ "$MT_flag" -ne 1 && "$selection" == A[lL][lL] ]]
    then
        rm ../"$2"\_ALL_scoring_list.csv
        echo "There are no HT/MT structures in this data set"
    fi
fi

cd ../






figures/setup/protocol/trajectory_factory.py
#!/usr/bin/env python


r"""

====================
Example Data factory
====================

.. moduleauthor:: Antonia Mey <antonia.mey@fu-berlin.de>

"""


#imports
import numpy as np
from scipy.integrate import quad
import math
import os



#some helper functions/classes
class AssymetricDoubleWellPotential ( object ):
    r"""class for an assymetic double well potential
    """
    def __init__( self ):
        self.Z = []
        self.inner_edges = np.linspace(-0.4,4.2,20)
        self.n_bins = self.inner_edges.shape[0]+1
        self.bin_centers = np.zeros(self.n_bins)
        self.bin_centers[1:-1] = self.inner_edges[1:]-0.5*( self.inner_edges[1]-self.inner_edges[0] )
        self.bin_centers[0] = -0.7
        self.bin_centers[-1] = 4.5

    def energy( self, x ):
        return 2*(x-2)-6*(x-2)**2 + (x-2)**4
    def gradient( self, x ):
        return 4*x**3-24*x**2+36*x-6
    def integ( self , x , kT ):
        return np.exp( - self.energy( x )/kT )
    def get_partition_function ( self, kT ):
        for T in kT:
            self.Z.append( quad(self.integ, -100.0, 100.0, args=(T,))[0] )
        self.Z = np.array( self.Z )
        return self.Z

class HarmonicRestraint( object ):
    r"""class for harmonic restraints used in US
    """

    def __init__( self, r0, k ):
        self.r0 = r0
        self.k = k
    def energy( self, r):
        return 0.5 * self.k * ( r - self.r0 )**2
    def gradient( self, r ):
        return self.k*(r-self.r0)


class BrownianIntegrator( object ):
    r"""
    class that does brownian dynamics integration
    """
    def __init__( self, potential, dt, beta=1.0, mass=1.0, damping=1.0 ):
        r""" constructor function
        
        Parameters
        ----------
        potential : FoldingPotential object
            object that contains all the information about the potential in which the particle svolves
        dt : double
            timestep of the integrator
        beta : double
            inverse temperature of the simulation Default = 1
        mass : double
            mass of the particle, Default = 1.0  
        damping : double
            damping constant, Default = 1.0
        """
        self.potential = potential
        # store parameters
        self.dt = dt
        self.beta = beta
        self.mass = mass
        self.damping = damping
        # compute coefficients
        self.coeff_A = dt / ( mass * damping )
        self.coeff_B = np.sqrt( 2.0 * dt / ( beta * mass * damping ) )
        self.x = None
        self.t_index = None #thermodynamic index
    def step( self, restraint = None ):
        r"""function that carries out a single integration step
        """
        gradient = self.potential.gradient( self.x )
        if None != restraint:
            gradient += restraint.gradient( self.x )
        self.x = self.x - self.coeff_A * gradient + self.coeff_B * np.random.normal( size=self.n_dim )
        pos = self.x[0]
        if self.n_dim > 1 :
            pos = np.linalg.norm(self.x)
        return np.array( [ pos , self.t_index, self.potential.energy( pos ) ] )

    def set_temperature( self, beta ):
        r"""function that sets a new inverse temperature
        Parameters
        ----------
        beta : double
            inverse temperature beta
        """
        self.beta = beta
        self.coeff_B = np.sqrt( 2.0 * self.dt / ( self.beta * self.mass * self.damping ) )
    def set_position( self, x ):
        r"""function that sets the current position of the integrator
        Parameters
        ----------
        x : double
            current position of the particle
        """
        self.x = x
        self.n_dim = x.shape[0]
    def set_t_index( self, t ):
        r"""function that sets the temperature index, should there be more than one
        Parameters
        ----------
        t : int
            temperature index in the temperature array
        """
        self.t_index = t

class STReplica( object ):
    def __init__( self,Z, integrator, kT ):
        r""" Constructor function
        Parameters
        ----------

        Z : double array
            containing normalisation factors used for weighting different temperatures
        integrator : BrownianIntegrator object
            the actual simulation
        kT : double array
            array contianing the possible temperatures at which sampling occurs
        """
        self.Z = Z
        self.integrator = integrator
        self.kT = kT
        self.trajectory = []
    def run( self, nsteps=1 ):
        r""" function that runs the replica exchange simulation
        Parameters
        ----------
        nsteps : int
            number of steps the integrator should compute between exchanges
            Default=1
        """
        for i in range( nsteps ):
            self.trajectory.append(self.integrator.step())

    def change_temperature( self ):
        r"""Metropolis hastings steps that allows to change between two randomly chosen temperatures
        """
        r = np.random.randint( self.kT.shape[0] )
        beta_new = 1.0/self.kT[r]
        beta_old = self.integrator.beta
        beta_int = np.where( self.kT == 1.0/beta_old )
        deltaG = -np.log ( self.integrator.potential.Z[r] ) + np.log( self.integrator.potential.Z[beta_int] )
        enExp = -self.trajectory[-1][2]*(beta_new-beta_old)
        exponent = enExp + deltaG
        
        if ( exponent >= 0 ) or ( np.random.random()< np.exp(exponent) ):
            self.integrator.set_temperature(1.0/self.kT[r])
            self.integrator.set_t_index(r)
            #print "New temperature is: "+str(self.integrator.beta)


class USReplica( object ):
    r""" class descriptor
    """
    def __init__( self,integrator, RS ):
        self.integrator = integrator
        self.RS = RS
        self.trajectory = []
        self.n_therm_states = len(RS)
        self.ndim=1

    def run( self, nsteps ):
        count = 0
        for r in self.RS:
            r_traj = []
            self.integrator.set_position(np.ones(self.ndim)*r.r0/np.sqrt(self.ndim))
            self.integrator.t_index = count
            for i in range( nsteps ):
                int_return = self.integrator.step(r)
                bias = self.calculate_bias(self.integrator.x)
                r_traj.append(np.hstack((int_return,bias)))
            count = count+1
            self.trajectory.append(r_traj)
        print ('US simulation completed sucessfully!')

    def calculate_bias( self, pos ):
        bias = np.zeros(self.n_therm_states)
        for b in range(self.n_therm_states):
            bias[b]=self.RS[b].energy(pos)
        return bias


def discretize( x, inner_edges ):
    if x<inner_edges[0]:
        return 0
    if x>=inner_edges[-1]:
        return inner_edges.shape[0]
    for i in range( inner_edges.shape[0]-1 ):
        if ( inner_edges[i]<=x) and ( x< inner_edges[i+1] ):
            return i+1

#the core run functions




def run_st_simulation():
    N_EXCHANGES=2000
    #checking if directories for writing exist
    directory="ST/"
    if not os.path.exists(directory):
        os.makedirs(directory)

    kT = np.array([ 2.0, 4.0, 7.0, 10.0, 15.0])
    n_therm_states = len(kT)
    initial_x = np.array([3.5])
    initial_t = 0

    dwp = AssymetricDoubleWellPotential()
    Z = dwp.get_partition_function(kT)
    integrator = BrownianIntegrator( dwp, 0.005, 1.0/kT[initial_t], 1.0, 1.0 )
    integrator.set_t_index(initial_t)
    integrator.set_position(initial_x)
    replica = STReplica(Z, integrator, kT )
    for i in range(N_EXCHANGES):
       replica.run(100)
       replica.change_temperature()
    traj = np.array(replica.trajectory)

    n_traj_frames =  np.shape(replica.trajectory)[0] 
    fh = open( directory+"Traj.dat", 'w' )
    for t in range( n_traj_frames ):
       fh.write( "%6d %6d %+.6e" % ( discretize( traj[t,0], dwp.inner_edges ), traj[t,1], traj[t,2] / kT[traj[t,1]] ) )
       fh.write( "\n" )
    fh.close()
    np.savetxt(directory+"kT.dat", kT)

    #constructing wham file
    # -> themodynamic state
    # |
    # v markov state
    #we need a target temperature \beta_0
    target = 0
    wham_f = os.path.join(directory,"b_K_i.dat")
    fh = open(wham_f, 'w')
    for c in range( dwp.bin_centers.shape[0] ):
       for i in range( n_therm_states ):
           fh.write( " %+.6e" % ( dwp.energy( dwp.bin_centers[c] ) * (1.0/kT[i]-1.0/kT[target] ) ) )
       fh.write( "\n" )
    fh.close()

    #exact probability distribution for comparison
    fh = open( directory+"exact.dat", 'w' )
    for c in range( dwp.bin_centers.shape[0] ):
       p=np.exp(-dwp.energy( dwp.bin_centers[c] )* 1.0/kT[target])
       fh.write( "%4f %+.6e" % ( dwp.bin_centers[c] , p ) )
       fh.write("\n")
    fh.close()



def run_us_simulation():
    #checking if directories for writing exist
    directory="US/"
    if not os.path.exists(directory):
       os.makedirs(directory)

    #setting the simulation temperature
    kT = np.array( [1.0] )
    nsteps=1000

    dwp = AssymetricDoubleWellPotential()
    Z = dwp.get_partition_function(kT)
    #exact probability distribution for comparison
    e_file = os.path.join(directory,"exact.dat")
    fh = open( e_file, 'w' )
    for c in range( dwp.bin_centers.shape[0] ):
       p=np.exp(-dwp.energy( dwp.bin_centers[c] )* 1.0/kT[0])
       fh.write( "%4f %+.6e" % ( dwp.bin_centers[c] , p ) )
       fh.write("\n")
    fh.close()

    integrator = BrownianIntegrator( dwp, 0.005, 1.0/kT[0], 1.0, 1.0) 
    restraints_pos = np.linspace(-0.9,4.7,30)
    restraint_k = np.ones(30)*90
    #restraints_pos = np.linspace(-1.8,1.8,16)
    n_therm_states = restraints_pos.shape[0]


    restraints = []
    for i in range(restraints_pos.shape[0]):
       restraints.append(HarmonicRestraint(restraints_pos[i],restraint_k[i]))
    replica = USReplica(integrator, restraints)
    replica.run(nsteps)
    for r in range(restraints_pos.shape[0]):
       traj = np.array(replica.trajectory[r])
       n_traj_frames = traj.shape[0]
       r_file = os.path.join(directory,"Traj"+str(r)+".dat")
       fh =open(r_file, 'w')
       for t in range( n_traj_frames ):
           fh.write( "%6d %6d " % (discretize( traj[t,0], dwp.inner_edges ) , traj[t,1] ) )
           for j in range(restraints_pos.shape[0]):
               fh.write("%+.6e " % (traj[t,3+j]/kT[0]))
           fh.write("\n")
       fh.close()

    wham_f = os.path.join(directory,"b_K_i.dat")
    fh = open(wham_f, 'w')
    for c in range( dwp.bin_centers.shape[0] ):
       for i in range( n_therm_states ):
           fh.write( " %+.8e" % ( restraints[i].energy( dwp.bin_centers[c] ) / kT[0] ) )
       fh.write( "\n" )
    fh.close()



examples/Confidence_intervals_RMSE_tau/plot.py
import matplotlib.pyplot as plt
import pandas as pd
import sys
from matplotlib import cm
from matplotlib import colors
import colorsys

def scale_lightness(rgb, scale_l):
    # convert rgb to hls
    h, l, s = colorsys.rgb_to_hls(*rgb)
    # manipulate h, l, s values and return as rgb
    return colorsys.hls_to_rgb(h, min(1, l * scale_l), s = s)

cmap = cm.get_cmap('hsv')
rgba = cmap(0.5)
all_results = pd.read_csv(sys.argv[1])
NN = all_results["N"]
lightcolors = [ scale_lightness(x,1.75) for x in [colors.ColorConverter.to_rgb(cmap(0.7)),colors.ColorConverter.to_rgb(cmap(0.8)),colors.ColorConverter.to_rgb(cmap(0.9))]]
fig, (ax1,ax2) = plt.subplots(2,1,figsize=(4.8,7))
l1 = ax1.plot(NN,all_results["mean_tau_0.5"],color=cmap(0.7),label=r'0.5')[0]
ax1.fill_between(NN,all_results["low_tau_0.5"],all_results["high_tau_0.5"],color=lightcolors[0])
l2 = ax1.plot(NN,all_results["mean_tau_1.0"],color=cmap(0.8))[0]
ax1.fill_between(NN,all_results["low_tau_1.0"],all_results["high_tau_1.0"],color=lightcolors[1])
l3 = ax1.plot(NN,all_results["mean_tau_1.5"],color=cmap(0.9))[0]
ax1.fill_between(NN,all_results["low_tau_1.5"],all_results["high_tau_1.5"],color=lightcolors[2])
ax1.set_xlabel('N')
ax1.set_ylabel(r'Kendall $\tau$')
ax2.plot(NN,all_results["mean_rmse_0.5"],color=cmap(0.7))
#plt.plot(NN,all_results[:,1],color='black')
#plt.plot(NN,all_results[:,2],color='black')
ax2.fill_between(NN,all_results["low_rmse_0.5"],all_results["high_rmse_0.5"],color=lightcolors[0])
ax2.plot(NN,all_results["mean_rmse_1.0"],color=cmap(0.8))
ax2.fill_between(NN,all_results["low_rmse_1.0"],all_results["high_rmse_1.0"],color=lightcolors[1])
ax2.plot(NN,all_results["mean_rmse_1.5"],color=cmap(0.9))
ax2.fill_between(NN,all_results["low_rmse_1.5"],all_results["high_rmse_1.5"],color=lightcolors[2])
ax2.set_xlabel('N')
ax2.set_ylabel('RMSE [kcal/mol]')
#plt.text(0.1,1.0,r"$\sigma = 1.0$",rotation=45,verticalalignment='center')
line_labels = [r'$\sigma=0.5$',r'$\sigma=1.0$',r'$\sigma=1.5$']
fig.legend([l1, l2, l3],     # The line objects
           labels=line_labels,   # The labels for each line
           loc=(0.6,0.6),   # Position of legend
           borderaxespad=0.1,    # Small spacing around legend box
           #title="Legend Title"  # Title for the legend
           )
#plt.subplots_adjust(hspace=0.25)
plt.savefig('N-CI.pdf',bbox_inches='tight')


examples/Confidence_intervals_RMSE_tau/simulate.py
import numpy as np
import scipy.stats
import bootstrapped.bootstrap as bs

def RMSE_function(values):
  tmp = np.sqrt(np.mean(values*values,axis=1))
  return tmp


def tau_function(a):
  s = a.shape
  tmp = np.array([])
  if len(s) == 2:
    tmp = scipy.stats.kendalltau(a[:,0],a[:,1])
  else:
    for i in range(s[0]):
      x = scipy.stats.kendalltau(a[i,:,0],a[i,:,1])[0]
      if x == np.nan:
        tmp = np.append(tmp,0.0)
      else:
        tmp = np.append(tmp,x)

  return tmp


def eval_stats(exp,pred,N):
  error = np.abs(exp-pred)
  rmse_bootstrap_dist = bs.bootstrap(np.reshape(error,(N,-1)), stat_func=RMSE_function,num_iterations=1000, alpha =0.05,is_pivotal=True,return_distribution=True)
  rmse = np.sqrt(np.mean(error**2))
  tau = scipy.stats.kendalltau(exp,pred)[0]
  tau_bootstrap_dist = bs.bootstrap(np.reshape(np.array(list(zip(exp,pred))),(N,-1)), stat_func=tau_function,num_iterations=1000, alpha =0.05,is_pivotal=True,return_distribution=True)
  tmp1 = [rmse,np.percentile(rmse_bootstrap_dist,2.5),np.percentile(rmse_bootstrap_dist,97.5),tau,np.percentile(tau_bootstrap_dist,2.5),np.percentile(tau_bootstrap_dist,97.5)]
  return tmp1


#generate toy data in interval [-12:-5]
all_results = []
NN = [10,25,35,50,75,100,200]
for N in NN:
#  simulate experimental data as uniform distribution
  data = [ np.random.uniform(low=-12,high=-5,size=N) for x in range(1000)]
  tmp = []
  for i in data:
    #simulated predicted data with gaussian error and calculate statistics for datasets
    tmpresults = []
    for sigma in [0.5,1.0,1.5,2.0]:
      i_pred = i + np.random.normal(0,sigma,N)
      results = eval_stats(i,i_pred,N)
      tmpresults.append(results)

    tmp.append(tmpresults)
    
  
  tmp = np.array(tmp)
  tmp = np.reshape(tmp,(1000,24))
  all_results.append(np.mean(tmp,axis=0))

all_results = np.array(all_results)
all_results = np.reshape(all_results,(-1,24))
import pandas as pd
df = pd.DataFrame(all_results,columns=['mean_rmse_0.5','low_rmse_0.5','high_rmse_0.5','mean_tau_0.5','low_tau_0.5','high_tau_0.5','mean_rmse_1.0','low_rmse_1.0','high_rmse_1.0','mean_tau_1.0','low_tau_1.0','high_tau_1.0','mean_rmse_1.5','low_rmse_1.5','high_rmse_1.5','mean_tau_1.5','low_tau_1.5','high_tau_1.5','mean_rmse_2.0','low_rmse_2.0','high_rmse_2.0','mean_tau_2.0','low_tau_2.0','high_tau_2.0'])
df['N']= NN
df.to_csv('results_N_CI_error.csv')



examples/structure_analysis/get_selected_ligands_for_rocs.sh
#!/bin/bash


# Test for input command line arguments
if [ $# -ne 4 ]; then
    echo "Please supply: <directory path> <target name> <quality> <irid/dpi cutoff>"
    exit
fi

if [ ! -d "$1"/prep/"$3" ]
then
    mkdir "$1"/prep/"$3"
fi

if [ ! -d "$1"/prep/"$3"/ligands ]
then
    mkdir "$1"/prep/"$3"/ligands
fi

# figure out what directory the files are in
#layer_count=$(tr -dc '/' <<<"$1" | wc -c)
#(( layer_count=layer_count+2 ))
#(( directory_count=layer_count-1 ))
#echo "$layer_count" "$directory_count"


for file in $(cat "$1"/"$2"\_"$3"\_score_stripped.csv)
do 
    test=$(echo "$file" | cut -d , -f 3)
    pdb_id=$(echo "$file" | cut -c 1-4)
    dpi=$(echo "$file" | cut -d , -f 6)
    irid_score=$(echo "$file" | cut -d , -f 7)
    pack=$(echo "$file" | cut -d , -f 8)
    if [[ "$4" == ALL ]]
    then
        if (( $(echo "$irid_score < $4" | bc -l) )) && [[ "$pack" == *false* ]]
        then
            cp "$1"/prep/"$pdb_id"*__ligand__*.oeb "$1"/prep/"$3"/ligands/
        fi
    else
       cp "$1"/prep/"$pdb_id"*__ligand__*.oeb "$1"/prep/"$3"/ligands/
    fi
done


examples/Confidence_intervals_RMSE_tau/plot_Helvetica.py
import matplotlib.pyplot as plt
import pandas as pd
import sys
from matplotlib import cm
from matplotlib import colors
import colorsys
from matplotlib import rc
font = {'size'   : 14}
rc('font', **font)
#rc('font',**{'family':'sans-serif','sans-serif':['Helvetica']})
## for Palatino and other serif fonts use:
#rc('font',**{'family':'serif','serif':['Palatino']})
#rc('text', usetex=True)

# change font
plt.rcParams['font.sans-serif'] = "Helvetica"
plt.rcParams['font.family'] = "sans-serif"

def scale_lightness(rgb, scale_l):
    # convert rgb to hls
    h, l, s = colorsys.rgb_to_hls(*rgb)
    # manipulate h, l, s values and return as rgb
    return colorsys.hls_to_rgb(h, min(1, l * scale_l), s = s)

cmap = cm.get_cmap('hsv')
rgba = cmap(0.5)
all_results = pd.read_csv(sys.argv[1])
NN = all_results["N"]
lightcolors = [ scale_lightness(x,1.75) for x in [colors.ColorConverter.to_rgb(cmap(0.7)),colors.ColorConverter.to_rgb(cmap(0.8)),colors.ColorConverter.to_rgb(cmap(0.9))]]
fig, (ax1,ax2) = plt.subplots(2,1,figsize=(4.8,7))
l1 = ax1.plot(NN,all_results["mean_tau_0.5"],color=cmap(0.7),label=r'0.5')[0]
ax1.fill_between(NN,all_results["low_tau_0.5"],all_results["high_tau_0.5"],color=lightcolors[0])
l2 = ax1.plot(NN,all_results["mean_tau_1.0"],color=cmap(0.8))[0]
ax1.fill_between(NN,all_results["low_tau_1.0"],all_results["high_tau_1.0"],color=lightcolors[1])
l3 = ax1.plot(NN,all_results["mean_tau_1.5"],color=cmap(0.9))[0]
ax1.fill_between(NN,all_results["low_tau_1.5"],all_results["high_tau_1.5"],color=lightcolors[2])
ax1.set_xlabel('N')
ax1.set_ylabel(r'Kendall $\tau$')
ax2.plot(NN,all_results["mean_rmse_0.5"],color=cmap(0.7))
#plt.plot(NN,all_results[:,1],color='black')
#plt.plot(NN,all_results[:,2],color='black')
ax2.fill_between(NN,all_results["low_rmse_0.5"],all_results["high_rmse_0.5"],color=lightcolors[0])
ax2.plot(NN,all_results["mean_rmse_1.0"],color=cmap(0.8))
ax2.fill_between(NN,all_results["low_rmse_1.0"],all_results["high_rmse_1.0"],color=lightcolors[1])
ax2.plot(NN,all_results["mean_rmse_1.5"],color=cmap(0.9))
ax2.fill_between(NN,all_results["low_rmse_1.5"],all_results["high_rmse_1.5"],color=lightcolors[2])
ax2.set_xlabel('N')
ax2.set_ylabel('RMSE [kcal/mol]')
#plt.text(0.1,1.0,r"$\sigma = 1.0$",rotation=45,verticalalignment='center')
line_labels = [r'$\sigma=0.5$',r'$\sigma=1.0$',r'$\sigma=1.5$']
fig.legend([l1, l2, l3],     # The line objects
           labels=line_labels,   # The labels for each line
           loc=(0.6,0.6),   # Position of legend
           borderaxespad=0.1,    # Small spacing around legend box
           #title="Legend Title"  # Title for the legend
           )
#plt.subplots_adjust(hspace=0.25)
plt.savefig('N-CI.pdf',bbox_inches='tight')
